hnn_core/gui/_viz_manager.py

```diff
@@ -0,0 +1,475 @@
+"""HNN GUI visualization management tool"""
+
+# Authors: Huzi Cheng <hzcheng15@icloud.com>
+
+import copy
+import io
+from functools import partial
+
+import matplotlib.pyplot as plt
+import numpy as np
+from IPython.display import display
+from ipywidgets import (Box, Button, Dropdown, FloatText, HBox, Label, Layout,
+                        Output, Tab, VBox, link)
+
+from hnn_core.gui._logging import logger
+from hnn_core.viz import plot_dipole
+
+
+class _TabWithNoneIndex(Tab):
+    def _reset_selected_index(self):
+        """Disable the default tab behavior that reset selected_index to 0."""
+        # if there are no tabs, then none should be selected
+        num_children = len(self.children)
+        if num_children == 0:
+            self.selected_index = None
+
+        # if there are tabs, but none is selected, select the first one
+        elif self.selected_index is None:
+            self.selected_index = None
+
+        # if there are tabs and a selection, but the selection is no longer
+        # valid, select the last tab.
+        elif num_children < self.selected_index:
+            self.selected_index = num_children - 1
+
+
+_dpi = 96
+_fig_placeholder = 'Run simulation to add figures here.'
+
+_plot_types = [
+    'current dipole',
+    'dipole@L2',
+    'dipole@L5',
+    'dipole@agg',
+    'input histogram',
+    'spikes',
+    'PSD',
+    'spectogram',
+    'network',
+    'layer-specific dipole',
+]
+
+_spectrogram_color_maps = [
+    "viridis",
+    "plasma",
+    "inferno",
+    "magma",
+    "cividis",
+]
+
+
+fig_templates = {
+    "2row x 1col (1:3)": {
+        "kwargs":
+        "gridspec_kw={\"height_ratios\":[1,3]}",
+        "mosaic": "00\n11",
+    },
+    "2row x 1col (1:1)": {
+        "kwargs":
+        "gridspec_kw={\"height_ratios\":[1,1]}",
+        "mosaic": "00\n11",
+    },
+    "1row x 2col (1:1)": {
+        "kwargs":
+        "gridspec_kw={\"height_ratios\":[1,1]}",
+        "mosaic": "01\n01",
+    },
+    "single figure": {
+        "kwargs": "",
+        "mosaic": "00\n00",
+    },
+    "2row x 2col (1:1)": {
+        "kwargs":
+        "gridspec_kw={\"height_ratios\":[1,1]}",
+        "mosaic": "01\n23",
+    },
+}
+
+
+def _idx2figname(idx):
+    return f"Figure {idx}"
+
+
+def _figname2idx(fname):
+    return int(fname.split(" ")[-1])
+
+
+def _update_ax(fig, ax, single_simulation, plot_type, plot_config):
+    """Refresh plots with simulation_data.
+
+    Parameters
+    ----------
+    fig: Figure
```

Comment:
nitpick but I think in most of the code base we're using an extra space with the colon for docstrings, as in: `variable : type`